### PR TITLE
✨Update to amp-3q-player

### PR DESCRIPTION
**Pull request by JuliusThms**
New config params for the player
- datasource
- datafield
- project
- key 
- timestamp

### CODE REVIEW

**Comment by JuliusThms** [comment]
Hi @alanorozco,

can you help with the owners-check?

Best, J

**Comment by JuliusThms** [comment]
@kristoferbaxter 

Hi, can you help own the owners-check.
What i have to do?

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 1]
```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
...
+    } else {
+      iframeSrc +=
+        this.dataId +
+        // Autoplay is handled by VideoManager
+        '?autoplay=false&amp=true';
+    }
 
-    this.iframe_ = iframe;
+    explicitParamsAttributes.forEach(explicitParam => {
+      const val = this.element.getAttribute(`data-${explicitParam}`);
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js]: You can just use `data-param-*` here like [`amp-video-iframe`](https://amp.dev/documentation/components/amp-video-iframe/). Use `getDataParamsFromAttributes` and you get that out of the box.

I don't think you need an allowlist for attributes if you explicitly set the `-param-` prefix.

File: extensions/amp-3q-player/validator-amp-3q-player.protoascii [hunk 2]
```diff
@@ -36,6 +36,26 @@ tags: {  # <amp-3q-player>
     name: "data-id"
     mandatory: true
   }
+  attrs: {
```
Comment (alanorozco) [inline extensions/amp-3q-player/validator-amp-3q-player.protoascii]: These are read as `data-` params, which don't need validation rules. Please undo this change.

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 3]
```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
...
         encodeURIComponent(dev().assertString(this.dataId)) +
         // Autoplay is handled by VideoManager
-        '?autoplay=false&amp=true'
-    );
+        '?autoplay=false&amp=true';
+    } else {
+      iframeSrc +=
+        this.dataId +
+        // Autoplay is handled by VideoManager
+        '?autoplay=false&amp=true';
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js]: Nit: Add this pair of params once when returning.

I would also prefer if you factor out a `getUrlPath()` method that returns the path value rather than modifying a string.

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 4]
```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
...
+        iframeSrc = addParamToUrl(iframeSrc, explicitParam, val);
+      }
+    });
+
+    return iframeSrc;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.iframe_ = this.generateIframeSrc();
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js]: Does this work? `iframeSrc` generates a string, whereas an `<iframe>` was previously stored from `createFrameFor`.

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 5]
```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
     Services.videoManagerForDoc(el).register(this);
   }
 
-  /** @override */
-  layoutCallback() {
-    const iframe = createFrameFor(
-      this,
-      'https://playout.3qsdn.com/' +
+  /** private */
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js]: This should be `/** @private */`, and the method should be suffixed with an underscore: `generateIframeSrc_()`

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 6]
```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
...
+        this.element.getAttribute(`data-datafield`) +
+        '/' +
         encodeURIComponent(dev().assertString(this.dataId)) +
         // Autoplay is handled by VideoManager
-        '?autoplay=false&amp=true'
-    );
+        '?autoplay=false&amp=true';
+    } else {
+      iframeSrc +=
+        this.dataId +
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js]: Shouldn't this be `URI`-encoded too? 

(Define variable once above the condition.)

File: extensions/amp-3q-player/0.1/amp-3q-player.js (hunk 1 above)
Comment (JuliusThms) [reply extensions/amp-3q-player/0.1/amp-3q-player.js]: The url structure changes with some params, i'll think about it.

File: extensions/amp-3q-player/validator-amp-3q-player.protoascii (hunk 2 above)
Comment (JuliusThms) [reply extensions/amp-3q-player/validator-amp-3q-player.protoascii]: done

File: extensions/amp-3q-player/0.1/amp-3q-player.js (hunk 3 above)
Comment (JuliusThms) [reply extensions/amp-3q-player/0.1/amp-3q-player.js]: done!

File: extensions/amp-3q-player/0.1/amp-3q-player.js (hunk 4 above)
Comment (JuliusThms) [reply extensions/amp-3q-player/0.1/amp-3q-player.js]: You we're right, corrected!

**Comment by JuliusThms** [comment]
@alanorozco 

I updated the code, can you review?

**Comment by JuliusThms** [comment]
@alanorozco 

Can you review, this changes needs to be merged soon...

**Comment by JuliusThms** [comment]
@alanorozco Can you check that? thanks!

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 7]
```diff
@@ -86,15 +87,39 @@ class Amp3QPlayer extends AMP.BaseElement {
...
+  /** private */
+  generateIframeSrc_() {
+    const explicitParamsAttributes = ['key', 'timestamp', 'controls'];
+
+    let iframeSrc = 'https://playout.3qsdn.com/';
+
+    if (this.element.getAttribute(`data-datasource`)) {
+      iframeSrc +=
+        'config_by_metadata/' +
+        this.element.getAttribute(`data-project`) +
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js:99]: These should be URI-encoded.

File: extensions/amp-3q-player/0.1/amp-3q-player.js [hunk 8]
```diff
@@ -86,15 +87,39 @@ class Amp3QPlayer extends AMP.BaseElement {
     Services.videoManagerForDoc(el).register(this);
   }
 
+  /** private */
```
Comment (alanorozco) [inline extensions/amp-3q-player/0.1/amp-3q-player.js:90]:
```suggestion
  /** @private */
```

**Comment by mrjoro** [comment]
Is this PR still active?

**Comment by JuliusThms** [comment]
@mrjoro Yes, ill finalize this shortly

**Comment by JuliusThms** [comment]
@alanorozco 
Requested changes are made.

**Comment by JuliusThms** [comment]
Can someone assist?
@mrjoro, @alanorozco

**Comment by alanorozco** [comment]
@JuliusThms Approved, pending a merge conflict. See comment above.

**Comment by JuliusThms** [comment]
> @JuliusThms Approved, pending a merge conflict. See comment above.

@alanorozco Thanks for your quick response! I resolved the conflict.

**Comment by alanorozco** [comment]
@JuliusThms Could you format the code? I can't merge due to [this lint error](https://travis-ci.org/github/ampproject/amphtml/jobs/710678759#L403).

**Comment by JuliusThms** [comment]
> @JuliusThms Could you format the code? I can't merge due to [this lint error](https://travis-ci.org/github/ampproject/amphtml/jobs/710678759#L403).

This is done, the tests are failing, can you have a look?

**Comment by JuliusThms** [comment]
> > @JuliusThms Could you format the code? I can't merge due to [this lint error](https://travis-ci.org/github/ampproject/amphtml/jobs/710678759#L403).
> 
> This is done, the tests are failing, can you have a look?

@alanorozco: Can you assist?

**Comment by alanorozco** [comment]
@JuliusThms Would you mind updating your branch with the latest master?

**Comment by JuliusThms** [comment]
@alanorozco I'll do so.

**Comment by JuliusThms** [comment]
@alanorozco I think you can merge.
